File: objectModel/TypeScript/Persistence/CdmFolder/EntityReferencePersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { CdmFolder } from '..';
import {
    CdmConstantEntityDefinition,
    CdmCorpusContext,
    CdmEntityDefinition,
    CdmEntityReference,
    cdmObjectType,
    CdmProjection,
    CdmTraitReferenceBase,
    copyOptions,
    resolveOptions
} from '../../internal';
import { cdmObjectRefPersistence } from './cdmObjectRefPersistence';
import { ProjectionPersistence } from './Projections/ProjectionPersistence';
import {
    ConstantEntity,
    Entity,
    EntityReferenceDefinition,
    Projection
} from './types';
import * as utils from './utils';

function isConstantEntity(object: Entity | ConstantEntity): object is ConstantEntity {
    return 'entityShape' in object;
}

export class EntityReferencePersistence extends cdmObjectRefPersistence {
    public static fromData(ctx: CdmCorpusContext, object: string | EntityReferenceDefinition | Projection): CdmEntityReference {
        if (!object) {
            return;
        }

        let simpleReference: boolean = true;
        let entity: string | CdmEntityDefinition | CdmConstantEntityDefinition | CdmProjection;
        let appliedTraits: CdmTraitReferenceBase[];
        
        if (typeof (object) === 'string') {
            entity = object;
        } else {
            simpleReference = false;
            entity = this.getEntityReference(ctx, object);
        }

        const entityReference: CdmEntityReference = ctx.corpus.MakeRef(cdmObjectType.entityRef, entity, simpleReference);

        if (typeof (object) !== 'string' && 'appliedTraits' in object) {
            appliedTraits = utils.createTraitReferenceArray(ctx, object.appliedTraits);
        }
        if (appliedTraits) {
            utils.addArrayToCdmCollection<CdmTraitReferenceBase>(entityReference.appliedTraits, appliedTraits);
        }

        return entityReference;
    }

    public static toData(instance: CdmEntityReference, resOpt: resolveOptions, options: copyOptions): any {
        if (instance.explicitReference !== undefined && instance.explicitReference instanceof CdmProjection) {
            return ProjectionPersistence.toData(instance.explicitReference as CdmProjection, resOpt, options) as Projection;
        }
        else {
            return cdmObjectRefPersistence.toData(instance, resOpt, options);
        }
    }

    private static getEntityReference(ctx: CdmCorpusContext, object: any): string | CdmEntityDefinition | CdmConstantEntityDefinition | CdmProjection {
        let entity: any = undefined;
        if ('entityReference' in object) {
            if (typeof (object.entityReference) === 'string') {
                entity = object.entityReference;
            } else if (isConstantEntity(object.entityReference)) {
                entity = CdmFolder.ConstantEntityPersistence.fromData(ctx, object.entityReference);
            } else {
                entity = CdmFolder.EntityPersistence.fromData(ctx, object.entityReference);
            }
        } else if ('source' in object || 'operations' in object) {
            entity = ProjectionPersistence.fromData(ctx, object);
        }

        return entity;
    }
}

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/cdmfolder/EntityReferencePersistence.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder;

import com.fasterxml.jackson.databind.JsonNode;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmEntityReference;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmTraitReferenceBase;
import com.microsoft.commondatamodel.objectmodel.cdm.projections.CdmProjection;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.persistence.cdmfolder.projections.ProjectionPersistence;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;

import java.util.List;

public class EntityReferencePersistence {

    public static CdmEntityReference fromData(final CdmCorpusContext ctx, final JsonNode obj) {
        if (obj == null) {
            return null;
        }

        final Object entity;
        boolean simpleReference = true;

        if (obj.isValueNode()) {
            entity = obj.asText();
        } else {
            entity = getEntityReference(ctx, obj);
            simpleReference = false;
        }

        final CdmEntityReference entityReference = ctx.getCorpus().makeRef(CdmObjectType.EntityRef, entity, simpleReference);

        if (!(obj.isValueNode())) {
            Utils.addListToCdmCollection(entityReference.getAppliedTraits(),
                                         Utils.createTraitReferenceList(ctx, obj.get("appliedTraits")));
        }

        return entityReference;
    }

    public static Object toData(final CdmEntityReference instance, final ResolveOptions resOpt, final CopyOptions options) {
        if (instance.getExplicitReference() != null && instance.getExplicitReference() instanceof CdmProjection) {
            return ProjectionPersistence.toData((CdmProjection) instance.getExplicitReference(), resOpt, options);
        } else {
            return CdmObjectRefPersistence.toData(instance, resOpt, options);
        }
    }

    private static Object getEntityReference(final CdmCorpusContext ctx, final JsonNode obj) {
        Object entity = null;
        if (obj.get("entityReference") != null && obj.get("entityReference").isValueNode()) {
            entity = obj.get("entityReference");
        } else if (obj.get("entityReference") != null && obj.get("entityReference").get("entityShape") != null) {
            entity = ConstantEntityPersistence.fromData(ctx, obj.get("entityReference"));
        } else if (obj.get("source") != null || obj.get("operations") != null) {
            entity = ProjectionPersistence.fromData(ctx, obj);
        } else {
            entity = EntityPersistence.fromData(ctx, obj.get("entityReference"));
        }

        return entity;
    }
}
